### PR TITLE
Context

### PR DESCRIPTION
**Pull request by ezolenko**
This includes standing PostCrafter's PR (sorry, it was in the way :))

Also:
- switch to `run(context)`
- updated test bundle (unhooked to anything though, but runs in my kernel!)
- optional rootImage/startContext because it is nice to be able to say:

```typescript
import { bundle as TestBundle } from "./posis/test.bundle";

posisKernel.loadBundle(TestBundle);

if (TestBundle.rootImageName)
	posisKernel.startRootProcess(TestBundle.rootImageName, TestBundle.defaultStartContext);
```

### CODE REVIEW

File: src/core/IPosisProcess.d.ts [hunk 1]
```diff
@@ -1,8 +1,15 @@
...
     id: PosisPID; // ID
     parentId: PosisPID; // Parent ID
     log: IPosisLogger; // Logger 
-    run(): void; // main function
+    queryPosisInterface<T extends keyof PosisInterfaces>(interfaceId: T): PosisInterfaces[T] | undefined;
+}
+
+interface IPosisProcess {
+    // Main function, implement all process logic here. 
+    // Bundle: Don't write to context object, host will likely freeze it anyway.
```
Comment (RiftLurker) [inline src/core/IPosisProcess.d.ts]: We could just mark the properties of `IPosisProcessContext` readonly.

File: src/core/IPosisBundle.d.ts [hunk 2]
```diff
@@ -0,0 +1,9 @@
+// Bundle for programs that are logically grouped
+interface IPosisBundle {
+	install(registry: IPosisProcessRegistry): void;
+
+	// image name of root process in the bundle, if any
+	rootImageName?: string;
+	// default start context for root process, if any
+	defaultStartContext?: any;
```
Comment (RiftLurker) [inline src/core/IPosisBundle.d.ts]: I realize we had a discussion in slack already, but I still don't think we need the `defaultStartContext`.

File: README.md [hunk 3]
```diff
@@ -13,19 +13,27 @@ Live editing [here](https://hackmd.io/GwBhBNgdgJgQwLQEYCcoEBYDMAjAxgnOOIjABwQBmM
...
+	// name your processes' image names with initials preceding, like ANI/MyCoolPosisProgram (but the actual class name can be whatever you want)
+	// if your bundle consists of several programs you can pretend that we have a VFS: "ANI/MyBundle/BundledProgram1"
+	register(imageName: string, constructor: new () => IPosisProcess): boolean;
+}
+```
+
+```typescript
+// Bundle for programs that are logically grouped
+interface IPosisBundle {
+	install(registry: IPosisProcessRegistry): void;
```
Comment (RiftLurker) [inline README.md]: I know it's an annoying task, but make sure to update the README as well.

File: src/core/IPosisBundle.d.ts (hunk 2 above)
Comment (anisoptera) [reply src/core/IPosisBundle.d.ts]: Yeah, I'm not a huge fan of this either. If you're looking to initialize your variables (just for your root process, too) then it seems fine to have to check "isInitialized" or something. Part of your root process's job can be to set up the environment, and making that explicit is good.

I also just don't see how I ever want to start a bundle without knowing the starting parameters.

File: examples/bundles/test.bundle.ts [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+}
+
+class POSISTest_BaseProcess implements IPosisProcess
+{
+	public testName = "POSIS base:";
+	public static ImageName = "POSISTest/PosisBaseTestProcess";
+
+	constructor(private context: IPosisProcessContext)
+	{		
+	}
```
Comment (AlinaNova21) [inline examples/bundles/test.bundle.ts:20]: should probably set `this.context` here, also, `this.context` isn't defined

File: src/extensions/IPosisSpawnExtension.d.ts [hunk 5]
```diff
@@ -17,8 +17,8 @@ interface IPosisSpawnExtension {
     spawnCreep(rooms: string[], body: string[][], memory?: any, opts?: { priority?: number }): string;
```
Comment (AlinaNova21) [inline src/extensions/IPosisSpawnExtension.d.ts:17]: remove `memory` parameter, if a process really needs to set it, they can do it on the returned creep or through Memory.creeps (Although both ways are heavily discouraged)

**Comment by AlinaNova21** [comment]
Other than those two things, looks good to me, although I still see no use whatsoever in rootImageName

File: examples/bundles/test.bundle.ts (hunk 4 above)
Comment (ezolenko) [reply examples/bundles/test.bundle.ts:20]: `private context: IPosisProcessContext` creates private prop on `this` automatically via typescript magic

File: src/extensions/IPosisSpawnExtension.d.ts (hunk 5 above)
Comment (ezolenko) [reply src/extensions/IPosisSpawnExtension.d.ts:17]: That's out of scope of this PR

File: examples/bundles/test.bundle.ts (hunk 4 above)
Comment (AlinaNova21) [reply examples/bundles/test.bundle.ts:20]: ahh, sneaky TS magic :)

**Pull request merged**